42 fast pattern operation with distributed command sequence

masato-fuk opened the pull request:
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- format added code and tests by `black -l 120 <filename>`
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
 
Then, please fill in below:

**Context (if applicable):**

na

**Description of the change:**

Implemented `LocalPattern` and `CommandNode`. With these new classes, `standardize` and `shift_signals` have been significantly accelerated.
Default backend for pattern operations is changed to `LocalPattern` class.

**Related issue:**

#42 

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

shinich1 commented:
could you add benchmarking results (plots) to this PR, briefly explaining what benchmarking was done?

shinich1 commented on tests/test_pattern.py:
```diff
@@ -193,5 +202,107 @@ def swap(circuit, a, b):
+        state_p = pattern.simulate_pattern()
+        state_ref = circuit.simulate_statevector()
+        np.testing.assert_almost_equal(np.abs(np.dot(state_p.flatten().conjugate(), state_ref.flatten())), 1)
+
```
what happens if someone tries normal standardization first, and then local standardization by mistake? or vice versa? or if mixed with signalshifting, e.g. standardize -> signalshift -> standardize? it's edge case but might be worth considering (e.g. throwing warning if we know they cause problem)

shinich1 commented:
Also - do we know local standardization work with 'opt' transpiration? sorry for adding so many cases but it's worth adding in test (if not working and it's a pain to fix now, we can let it raise warning so that users know it's not desired way to code)

masato-fuk commented on graphix/pattern.py:
```diff
@@ -1080,6 +1171,350 @@ def to_qasm3(self, filename):
+        Returns
+        -------
+        signal_destination : set
+            measurement results of each node propagate to the nodes specified by 'signal_distination'. Inversal set of dependent nodes.
```
I mean 'counterpart' or 'inversal mapping' here, instead of 'inversal set'. My word-choice wasn't good.

masato-fuk commented:
I modified local pattern to support opt-transpilation and added a test for it. Please check [bf1fb0d](https://github.com/TeamGraphix/graphix/pull/43/commits/bf1fb0d591b86259557b83abd2ff4d7665672565)

shinich1 commented:
@masato-fuk  Do you want me to review now?

masato-fuk commented:
I'm sorry for mistouching the `merge button` at [e3edf4d](https://github.com/TeamGraphix/graphix/pull/43/commits/e3edf4dc8f8caaf49b8562c24b5acf5fb10e941e).
But, I've completed all ToDos in this issue. So, would you review changes when you have a time?
Benchmark scores of the new methods will be added in this conversation when I finishied benchmark with opt-transpilation and localpattern. Perhaps tonight.

masato-fuk commented:
![pattern_operation_benchmark](https://user-images.githubusercontent.com/119649569/235032704-b69992f2-b3de-4a7a-9c19-056223911dfa.png)
Performance check of new methods(local pattern) compared with conventional methods(global pattern).
the opt-transpilation method is still valid for local pattern operations

masato-fuk commented:
![localpattern_std](https://user-images.githubusercontent.com/119649569/235036878-caab08f8-7e10-44a3-aa8d-128e58d35b16.png)
![localpattern_signal](https://user-images.githubusercontent.com/119649569/235036886-a6fef5e4-0fce-4602-9edb-b27fd778db0a.png)
The first graph is benchmark of local pattern standardization up to 50000 nodes.
The second graph is benchmark of local pattern signal shifting up tp the same size.
I'm not sure what happens around the several thousands of nodes region.

shinich1 commented on graphix/pattern.py:
```diff
@@ -1080,6 +1166,370 @@ def to_qasm3(self, filename):
+        return pattern
+
+
+def xor_combination_list(list1, list2):
```
Can we use python set operations?

masato-fuk replied:
I think it will be complicated to use set because it cannot distinguish one where a signal appears only once from where a signal appears more than once. (e.g. set treats the following as the same, [1]. [1, 1], [1, 1, 1] -> {1}).
Do you have any idea to solve the above problem?

shinich1 commented on graphix/pattern.py:
```diff
@@ -1080,6 +1166,370 @@ def to_qasm3(self, filename):
+        pattern : Pattern
+            standardized global pattern
+        """
+        pattern = Pattern(output_nodes=self.output_nodes)
```
Can we `assert` it’s standardised?

masato-fuk replied:
Sure. I will add a checker.

shinich1 commented:
@masato-fuk  please squash and merge, with commit message
`fast standardization method with localized command structure (#43)`

shinich1 commented:
@masato-fuk added `standardize_and_shift_signals` method. please check and squash and merge, if it looks ok.
